metaflow/client/core.py

```diff
@@ -1211,9 +1419,12 @@ def metadata_dict(self) -> Dict[str, str]:
             Dictionary mapping metadata name with value
         """
         # use the newest version of each key, hence sorting
-        return {
-            m.name: m.value for m in sorted(self.metadata, key=lambda m: m.created_at)
-        }
+        if self._metadata_dict is None:
+            self._metadata_dict = {
+                m.name: m.value
+                for m in sorted(self.metadata, key=lambda m: m.created_at)
+            }
+        return self._metadata_dict
```

Comment:
Note: this slightly changes the syntax since now if there is new metadata, the user won't get it. Should check if this impacts other operations. Or scope the caching to just the functions that need it.